src/client.ts

```diff
@@ -0,0 +1,235 @@
+
+function parseRequestBody<RequestBody>({
+	body,
+	requestBodySchema,
+	path,
+}: {
+	body: RequestBody
+	requestBodySchema?: ZodSchema<RequestBody>
+	path: string
+}): Either<ZodError, RequestBody> {
+	if (!body) {
+		return success(body)
+	}
+
+	if (!requestBodySchema) {
+		return success(body as RequestBody)
+	}
+
+	const result = requestBodySchema.safeParse(body)
+
+	if (!result.success) {
+		console.error({
+			path,
+			body,
+			error: result.error,
+		})
+		return failure(result.error)
+	}
+
+	return success(body)
+}
+
+function parseQueryParams<RequestQueryParams>({
+	queryParams,
+	queryParamsSchema,
+	path,
+}: {
+	queryParams: RequestQueryParams
+	queryParamsSchema?: ZodSchema<RequestQueryParams>
+	path: string
+}): Either<ZodError, string> {
+	if (!queryParams) {
+		return success('')
+	}
+
+	if (!queryParamsSchema) {
+		return success(`?${stringify(queryParams)}`)
+	}
+
+	const result = queryParamsSchema.safeParse(queryParams)
+
+	if (!result.success) {
+		console.error({
+			path,
+			queryParams,
+			error: result.error,
+		})
+		return failure(result.error)
+	}
+
+	return success(`?${stringify(queryParams)}`)
+}
+
+function parseResponseBody<ResponseBody>({
+	response,
+	responseBodySchema,
+	path,
+}: {
+	response: ResponseBody
+	responseBodySchema?: ZodSchema<ResponseBody>
+	path: string
+}): Either<ZodError, ResponseBody> {
+	if (!responseBodySchema) {
+		return success(response)
+	}
+
+	const result = responseBodySchema.safeParse(response)
+
+	if (!result.success) {
+		console.error({
+			path,
+			response,
+			error: result.error,
+		})
+
+		return failure(result.error)
+	}
+
+	return success(response)
+}
+
+async function sendResourceChange<
+	T extends WretchInstance,
+	ResponseBody,
+	RequestBody extends object | undefined = undefined,
+	RequestQueryParams extends object | undefined = undefined,
+>(
+	wretch: T,
+	method: 'post' | 'put' | 'patch',
+	params: ResourceChangeParams<RequestBody, ResponseBody, RequestQueryParams>,
+) {
+	const body = parseRequestBody({
+		body: params.body,
+		requestBodySchema: params.requestBodySchema,
+		path: params.path,
+	})
+
+	if (isFailure(body)) {
+		return Promise.reject(body.error)
+	}
+
+	const queryParams = parseQueryParams({
+		queryParams: params.queryParams,
+		queryParamsSchema: params.queryParamsSchema,
+		path: params.path,
+	})
+
+	if (isFailure(queryParams)) {
+		return Promise.reject(queryParams.error)
+	}
+
+	return (
+		wretch[method](body.result, `${params.path}${queryParams.result}`)
+			// HTTP 204 No Content response should return undefined body
+			.res((response) => (response.status === 204 ? undefined : response.json()))
+			.then((response) => {
+				const parsedResponse = parseResponseBody({
+					response: response as ResponseBody,
+					responseBodySchema: params.responseBodySchema,
+					path: params.path,
+				})
+
+				if (isFailure(parsedResponse)) {
+					return Promise.reject(parsedResponse.error)
+				}
+
+				return parsedResponse.result
+			})
```

Comment:
🔶 Not the latest version. This was changed yesterday in AP repo - and with it also tests.

Reply:
oooof, thanks. I must have not pulled the code today. will fix now

Reply:
fixed!